pkg/cmd/image/push.go

```diff
@@ -159,6 +160,16 @@ func Push(ctx context.Context, client *containerd.Client, rawRef string, options
 		if err != nil {
 			return err
 		}
+	case "notation":
+
+		if !options.GOptions.Experimental {
+			return fmt.Errorf("notation only work with enable experimental feature")
+		}
+
+		err = notationutil.SignNotation(rawRef, options.NotationKeyName)
+		if err != nil {
+			return err
+		}
```

Comment:
discussion: this seems quite repeatitive, we have this `switch` logic (and experimental check logic) in pull and push (and compose pull and push).

The function signature for sign (and verify) with cosign (and notation) looks the same. Probably this repeatitive logic can be abstracted into the `signutil` package, for example:

1. Have a `signutil.Sign(rawRef, signKey, options.Sign, options.Experimental` (and correspondingly `signutil.Verify`) that can handle those repeatitive logics. Then if we add a new tool (probably won't atm) we only need to change in 1 place; if one tool is out of experimental, we also only need to change 1 place.
2. Have a `SigningTool` type/interface that has `Sign`/`Verify`/`IsExperimental` funcs. Cosign/Notation will implement this interface. And we can have a factory somewhere to init an instance based on `Option.Sign`.

Reply:
Cosign and Notation actually have much different interface for sign and verify.
Cosign needs the private key file to be passed in `sign` while Notation needs the key name (it is the alias in notation specific signing-keys.json) to be passed.
Cosign needs the public key file to be passed in `verify` while Notation needs no keys in `verity`.

The issue https://github.com/containerd/nerdctl/issues/1974 has more details.

Seems the signing tools don't have a unique CLI spec, the future 3rd tools (if any) might have another different interfaces.

From my understanding, I guess your idea of both 1 and 2 require them having same interfaces? Let me know I'm understanding wrong.

Reply:
I see, thanks for the explanation. I'm just trying to figure out if it's possible to reduce duplicated logic (e.g., [here](https://github.com/containerd/nerdctl/blob/2880ddb5f1f13363b578cde5895c5f52cd68d229/pkg/cmd/compose/compose.go#L135), [here](https://github.com/containerd/nerdctl/blob/82467b106ff87a69b3e537b51f215a71ca697571/pkg/cmd/image/pull.go#L91) and [here](https://github.com/containerd/nerdctl/blob/5da1481d3342d75bf356469afeac2fb534363047/pkg/cmd/image/push.go#L152)) and potentially reduce future changes. I'm okay with the current format since we have only 2 tools.

Just for discussion, one possible refactor might be to put all sign/verify related fields (both cosign, notation) into a  `SignOptions`/`VerifyOptions`, so the related command options (e.g. `ImagePushOptions`) don't hold all individual sign/verify fields but a single `SignOptions` field.

Then we can have a dispatcher func like `signutil.Sign(opt SignOptions)` that has the switch-case/experimental-check logic (`opt` has all cosign/notation related fields).

Now, if we want to add a new tool:

1. Update one `SignOptions` struct (not all `PullOptions`/`PushOptions`/`RunOptions`/etc structs) (cmd parsing update still need to be done per cmd).
2. Update the dispatcher func `signtuil.Sign(opt SignOptions)` to add the `case` for new tool (instead of all the places mentioned above).

If one tool goes out of experimental:

1. Update the dispatcher func to remove the experimental-check (instead of removing in all the places mentioned above).

Reply:
Great. This sounds like the right refactoring direction to me.